### PR TITLE
[M3c][MetaScheduler] Update TuneContext, TaskScheduler & Search Strategy Design

### PR DESCRIPTION
**Pull request by zxybazh**
This PR is part of the stage M3c of the meta schedule project (#8473).

The architecture is re-designed by Junru and Xiyou. In this PR we modified the design of multiple classes and added post processings, mutators to the meta schedule framework. Also more concrete classes would be added later. We intend to make the interfaces more flexible yet user-friendly with updates in the PR, mostly in SearchStrategy class.

Unittests are also updated in this PR. Thanks to all co-authors for contributing!

<s>Note: The PR currenlt includes PR #9780 and will rebase once it's merged.</s>

Co-authored-by: Junru Shao <<junrushao1994@gmail.com>>
Co-authored-by: Bohan Hou <<32121147+spectrometerHBH@users.noreply.github.com>>
Co-authored-by: Ruihang Lai <<lairuihangdongdong@qq.com>>
Co-authored-by: Hongyi Jin <<3231950289@qq.com>>
Co-authored-by: Wuwei Lin <<wuwei@apache.org>>
Co-authored-by: Siyuan Feng <<Hzfengsy@sjtu.edu.cn>>

### CODE REVIEW

**Comment by zxybazh** [comment]
CC @comaniac @jcf94 @merrymercy

File: include/tvm/meta_schedule/tune_context.h [hunk 1]
```diff
@@ -41,34 +48,48 @@ class TuneContextNode : public runtime::Object {
   Optional<SearchStrategy> search_strategy;
   /*! \brief The schedule rules. */
   Array<ScheduleRule> sch_rules;
+  /*! \brief The postprocessors. */
+  Array<Postproc> postprocs;
+  /*! \brief The probability of using certain mutator. */
+  Map<Mutator, FloatImm> mutator_probs;
   /*! \brief The name of the tuning task. */
-  Optional<String> task_name;
+  String task_name;
```
Comment (junrushao1994) [inline include/tvm/meta_schedule/tune_context.h:56]: Why is it changed to non-optional?

File: include/tvm/meta_schedule/tune_context.h [hunk 2]
```diff
@@ -41,34 +48,48 @@ class TuneContextNode : public runtime::Object {
...
     v->Visit("sch_rules", &sch_rules);
+    v->Visit("postprocs", &postprocs);
+    v->Visit("mutator_probs", &mutator_probs);
     v->Visit("task_name", &task_name);
     v->Visit("rand_state", &rand_state);
     v->Visit("num_threads", &num_threads);
     v->Visit("is_stopped", &is_stopped);
+    v->Visit("builder_results", &builder_results);
     v->Visit("runner_futures", &runner_futures);
     v->Visit("measure_candidates", &measure_candidates);
```
Comment (junrushao1994) [inline include/tvm/meta_schedule/tune_context.h:87]: add `builder_results` and `runner_futures`

File: python/tvm/auto_scheduler/search_task.py [hunk 3]
```diff
@@ -543,7 +543,8 @@ def print_best(self, log_file, print_mode="schedule"):
         code: str
             The best schedule code in python API or CUDA source code
         """
-        inp, _ = load_best_record(log_file, self.workload_key)
+        inp, res = load_best_record(log_file, self.workload_key)
+        print("Best codes (ms):", [float(c) * 1000.0 for c in res.costs])
```
Comment (junrushao1994) [inline python/tvm/auto_scheduler/search_task.py:547]: remove this change

File: python/tvm/auto_scheduler/workload_registry.py [hunk 4]
```diff
@@ -194,7 +194,10 @@ def workload_key_to_tensors(workload_key):
     assert callable(value)
 
     args = deserialize_args(workload[1:])
-    return value(*args)
+    result = value(*args)
+    if isinstance(result, tuple):
+        result = list(result)
+    return result
```
Comment (junrushao1994) [inline python/tvm/auto_scheduler/workload_registry.py:200]: no need for this change

File: src/meta_schedule/search_strategy/replay_trace.cc [hunk 5]
```diff
@@ -17,27 +17,26 @@
  * under the License.
  */
 #include "../utils.h"
+#include "tvm/tir/schedule/schedule.h"
```
Comment (junrushao1994) [inline src/meta_schedule/search_strategy/replay_trace.cc:20]: let's fix this?

File: src/meta_schedule/utils.h [hunk 6]
```diff
@@ -233,6 +242,73 @@ inline std::string Concat(const Array<String>& strs, const std::string& delim) {
...
+  }
+
+ private:
+  struct Item {
+    Postproc postproc{nullptr};
+    std::atomic<int> fail_counter{0};
+  };
+
+  int n_;
+  Item* items_;
```
Comment (junrushao1994) [inline src/meta_schedule/utils.h:309]: doc?

File: include/tvm/meta_schedule/tune_context.h (hunk 2 above)
Comment (zxybazh) [reply include/tvm/meta_schedule/tune_context.h:87]: Do you mean remove?

File: include/tvm/meta_schedule/tune_context.h (hunk 1 above)
Comment (zxybazh) [reply include/tvm/meta_schedule/tune_context.h:56]: It should be optional, will fix.

**Review by junrushao1994 — APPROVED** [review]
LGTM

**Pull request merged**